Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/non_rigid_mesh_registration.h

```diff
@@ -0,0 +1,930 @@
+
+    cr.add_scalar_t_vector_t_vector_transpose(cov, *w, source_edge, moving_edge);
+    v.update_covariance_matrix(cov, rotations[index]);
+  }
+
+  cr.compute_close_rotation(cov, rotations[index]);
+}
+
+template <typename T>
+Eigen::Matrix<T, 3, 3> rot(T a, T b, T c) {
+  T ca = cos(a), cb = cos(b), cc = cos(c);
+  T sa = sin(a), sb = sin(b), sc = sin(c);
+  Eigen::Matrix<T, 3, 3> R;
+  R << cb * cc, cc* sa* sb - ca * sc, ca* cc* sb + sa * sc,
+    cb* sc, ca* cc + sa * sb * sc, ca* sb* sc - cc * sa,
+    -sb, cb* sa, ca* cb;
+
+  return R;
+}
+
+#endif
+} // namespace registration
+} // namespace internal
+
+/*!
+* \ingroup PMP_registration_grp
+*
+* \brief computes non-rigid transformation of a mesh onto a set of oriented points.
+*
+* A non-rigid ICP, iterative closest point, method based on
+* <A HREF="https://vgl.ict.usc.edu/Research/NonRigidRegistration/MODERN%20TECHNIQUES%20AND%20APPLICATIONS%20FOR%20REAL-TIME%20NON-RIGID%20REGISTRATION.pdf">a SIGGRAPH'16 Tutorial</A>.
+* The method uses a few `correspondences` between the `source` and the `target` for the rough alignment. The iterative closest point method
+* iteratively approaches the `target` by minimizing the distance between vertices of the `source` and points of the `target`.
+*
+* @note This function requires the \ref thirdpartyEigen library.
+*
+* @tparam TriangleMesh a model of `MutableFaceGraph`.
+* @tparam PointRange is a model of `ConstRange`. The value type of
+*   its iterator is the key type of the named parameter `point_map`.
+* @tparam VertexTranslationMap is a property map with `boost::graph_traits<TriangleMesh>::%vertex_descriptor`
+*   as key type and a \cgal Kernel `Vector_3` as value type.
+* @tparam VertexRotationMap is a property map with `boost::graph_traits<TriangleMesh>::%vertex_descriptor`
+*   as key type and a \cgal Kernel `Aff_transformation_3` as value type.
+* @tparam NamedParameters1 a sequence of \ref bgl_namedparameters "Named Parameters".
+* @tparam NamedParameters2 a sequence of \ref bgl_namedparameters "Named Parameters".
+*
+* @param source the triangle mesh to be mapped onto `target`.
+* @param target the target point set.
+* @param vtm a writable vertex property map of `source` to store the translation vector of the registration.
+* @param vrm a writable vertex property map of `source` to store the rotation part of the registration.
+* @param np1 an optional sequence of \ref bgl_namedparameters "Named Parameters 1" among the ones listed below.
+* @param np2 an optional sequence of \ref bgl_namedparameters "Named Parameters 2" providing a point_map and normal_map for the `PointRange` as listed below.
+*
+* \cgalNamedParamsBegin{Named Parameters 1}
+*   \cgalParamNBegin{number_of_iterations}
+*     \cgalParamDescription{the number of registration iterations using ICP, iterative closest point}
+*     \cgalParamType{unsigned int}
+*     \cgalParamDefault{`50`}
+*   \cgalParamNEnd
+*
+*   \cgalParamNBegin{point_to_plane_energy}
```

Comment:
providing a link to the section in the user manual where parameters are explained would be helpful. The range of possible values for w1, w2, w3 is also needed.